lib/std/os.zig

```diff
@@ -3121,13 +3121,21 @@ pub fn getsockoptError(sockfd: fd_t) ConnectError!void {
     }
 }
 
-pub fn waitpid(pid: i32, flags: u32) u32 {
+
+pub const WaitpidRet = struct {
+   pid: i32,
+   status: u32
+};
+
+pub fn waitpid(pid: i32, flags: u32) waitpid_ret {
     // TODO allow implicit pointer cast from *u32 to *c_uint ?
     const Status = if (builtin.link_libc) c_uint else u32;
+    const Return = if (builtin.link_libc) c_int else usize;
     var status: Status = undefined;
     while (true) {
-        switch (errno(system.waitpid(pid, &status, flags))) {
-            0 => return @bitCast(u32, status),
+        var ret: Return = system.waitpid(pid, &status, flags);
+        switch (errno(ret)) {
+            0 => return waitpid_ret{ .pid =  if (@TypeOf(ret) == c_int) @intCast(i32,ret) else  @intCast(i32, @truncate(u32, ret)), .status = @bitCast(u32, status) },
```

Comment:
Logic here should be something like this:
```zig
const R = @TypeOf(ret);
const R2 = meta.Int(isSigned(pid_t), R.bits);
const ret2: R2 = if (@boolToInt(isSigned(R)) ^ @boolToInt(isSigned(pid_t)) == 0) ret else @bitCast(R2, ret);
.pid = math.cast(pid_t, ret2) catch etc...
```

Reply:
intCast is probably fine here, but i do not know if its always signed/unsigned and on what oses, to be extra safe you can use more complicated version

Reply:
In what case would @intCast() not be fine?

Reply:
FWIW we use `@intCast(pid_t, rc)` in `os.fork()` already.

Reply:
On linux pid_t is i32, ret is usize, if MSB is set then intCast will fail

Reply:
on linux pids start at 0 and are allocated sequentially up to `/proc/sys/kernel/pid_max`. `pid_t` is signed only so that `fork()` may return -1 on failure afaik.

Reply:
afaik pid_t cannot be more than 32767 on the linuxen i normally use be it 32 or 64 bit.

Reply:
Good, anyway it will panic with a stack trace in debug mode if that ever happens